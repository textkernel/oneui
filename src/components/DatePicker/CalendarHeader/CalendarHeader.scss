.CalendarHeader {
    display: flex;
    justify-content: space-between;
    align-items: baseline;
    margin-bottom: var(--spacing-3x);
    color: var(--color-neutral-75);

    &__navButton {
        width: 33px;
        height: 33px;
        padding: var(--spacing-normal);
        border: transparent;
    }

    &__select {
        margin-left: var(--spacing-normal);
        background-color: var(--color-background);
        border: 0;
        border-bottom: 1px var(--color-neutral) solid;
        font-size: var(--font-size-large);

        &:focus {
            outline: none;
        }
    }
}
