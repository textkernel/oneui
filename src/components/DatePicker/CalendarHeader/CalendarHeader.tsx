import * as React from 'react';
import { ReactDatePickerCustomHeaderProps } from 'react-datepicker';
import { ImArrowLeft2, ImArrowRight2 } from 'react-icons/im';
import { bem } from '../../../utils';
import { Button } from '../../Buttons';
import { Text } from '../../Text';
import styles from './CalendarHeader.scss';

interface Props extends ReactDatePickerCustomHeaderProps {
    /** the locale in which the names of the month should be displayed. Format should match Intl definitions */
    locale?: string;
    /** the min and max year selectable by the user */
    yearsRange: [number, number];
}

const { block, elem } = bem('CalendarHeader', styles);

export const CalendarHeader: React.FC<Props> = (props) => {
    const {
        locale = 'en',
        yearsRange,
        date,
        decreaseMonth,
        prevMonthButtonDisabled,
        increaseMonth,
        nextMonthButtonDisabled,
        changeYear,
    } = props;

    const monthName = new Intl.DateTimeFormat(locale, { month: 'long' }).format(date);

    const [minYear, maxYear] = yearsRange;
    if (maxYear < minYear) {
        console.error(
            `CalendarHeader component has received invalid props.
            Minimum selectable year (${minYear}) is larger then maximum selectable year (${maxYear})`
        );
        return null;
    }

    const yearsAvailable = maxYear - minYear + 1;
    const years = [...new Array(yearsAvailable)].map((val, idx) => minYear + idx);

    return (
        <div {...block()}>
            <Button
                onClick={decreaseMonth}
                disabled={prevMonthButtonDisabled}
                {...elem('navButton')}
            >
                <ImArrowLeft2 />
            </Button>
            <div role="presentation">
                <Text size="large" inline>
                    {monthName}
                </Text>
                <select
                    {...elem('select')}
                    value={date.getFullYear()}
                    onChange={({ target: { value } }) => changeYear(parseInt(value, 10))}
                >
                    {years.map((option) => (
                        <option key={option} value={option}>
                            {option}
                        </option>
                    ))}
                </select>
            </div>
            <Button
                onClick={increaseMonth}
                disabled={nextMonthButtonDisabled}
                {...elem('navButton')}
            >
                <ImArrowRight2 />
            </Button>
        </div>
    );
};

CalendarHeader.displayName = 'CalendarHeader';
