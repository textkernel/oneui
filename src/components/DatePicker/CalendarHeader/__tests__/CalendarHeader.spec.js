import React from 'react';
import toJson from 'enzyme-to-json';
import { CalendarHeader } from '../CalendarHeader';

describe('CalendarHeader', () => {
    const YEARS = [2016, 2025];
    const decreaseMonth = jest.fn();
    const increaseMonth = jest.fn();
    const changeYear = jest.fn();
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <CalendarHeader
                yearsRange={YEARS}
                date={new Date('2021-12-31')}
                decreaseMonth={decreaseMonth}
                increaseMonth={increaseMonth}
                changeYear={changeYear}
            />
        );
    });

    it('should render correctly', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    it('should log an error and render nothing if min year is larger then max year', () => {
        // don't log to test output
        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementationOnce(() => {});
        wrapper.setProps({ yearsRange: [9, 2] });
        wrapper.update();

        expect(wrapper.html()).toBe(null);
        expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
        expect(consoleErrorSpy.mock.calls[0][0].replace(/\s/g, '')).toMatch(
            `CalendarHeader component has received invalid props.
        Minimum selectable year (9) is larger then maximum selectable year (2)`.replace(/\s/g, '')
        );
    });
    it('should fill in the years correctly', () => {
        const options = wrapper.find('option');
        // eslint-disable-next-line no-plusplus
        for (let year = YEARS[0]; year <= YEARS[1]; year++) {
            const idx = year - YEARS[0];
            expect(options.at(idx).prop('value')).toEqual(year);
        }
    });
    it('should display the month in the correct language', () => {
        expect(wrapper.find('Text').text()).toBe('December');
    });
    it('should call decreaseMonth when previous nav button is clicked', () => {
        wrapper.find('button').at(0).simulate('click');
        expect(decreaseMonth).toHaveBeenCalledTimes(1);
    });
    it('should call increaseMonth when next nav button is clicked', () => {
        wrapper.find('button').at(1).simulate('click');
        expect(increaseMonth).toHaveBeenCalledTimes(1);
    });
    it('should call changeYear when year is selected', () => {
        wrapper.find('select').simulate('change', { target: { value: 2019 } });
        expect(changeYear).toHaveBeenCalledTimes(1);
        expect(changeYear).toHaveBeenCalledWith(2019);
    });
});
