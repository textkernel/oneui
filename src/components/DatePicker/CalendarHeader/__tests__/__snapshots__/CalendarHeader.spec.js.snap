// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader should render correctly 1`] = `
<CalendarHeader
  changeYear={[MockFunction]}
  date={2021-12-31T00:00:00.000Z}
  decreaseMonth={[MockFunction]}
  increaseMonth={[MockFunction]}
  yearsRange={
    Array [
      2016,
      2025,
    ]
  }
>
  <div
    className="CalendarHeader"
  >
    <Button
      className="CalendarHeader__navButton"
      context="neutral"
      disabled={false}
      isBlock={false}
      isInline={false}
      onClick={[MockFunction]}
      size="normal"
      type="button"
    >
      <button
        className="Button Button--context_neutral CalendarHeader__navButton"
        disabled={false}
        onClick={[MockFunction]}
        type="button"
      >
        <ImArrowLeft2>
          <IconBase
            attr={
              Object {
                "version": "1.1",
                "viewBox": "0 0 16 16",
              }
            }
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.1"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.293 13.707l-5-5c-0.391-0.39-0.391-1.024 0-1.414l5-5c0.391-0.391 1.024-0.391 1.414 0s0.391 1.024 0 1.414l-3.293 3.293h9.586c0.552 0 1 0.448 1 1s-0.448 1-1 1h-9.586l3.293 3.293c0.195 0.195 0.293 0.451 0.293 0.707s-0.098 0.512-0.293 0.707c-0.391 0.391-1.024 0.391-1.414 0z"
                key="0"
              />
            </svg>
          </IconBase>
        </ImArrowLeft2>
      </button>
    </Button>
    <div
      role="presentation"
    >
      <Text
        context="default"
        inline={true}
        size="large"
      >
        <span
          className="Text--size_large"
        >
          December
        </span>
      </Text>
      <select
        className="CalendarHeader__select"
        onChange={[Function]}
        value={2021}
      >
        <option
          key="2016"
          value={2016}
        >
          2016
        </option>
        <option
          key="2017"
          value={2017}
        >
          2017
        </option>
        <option
          key="2018"
          value={2018}
        >
          2018
        </option>
        <option
          key="2019"
          value={2019}
        >
          2019
        </option>
        <option
          key="2020"
          value={2020}
        >
          2020
        </option>
        <option
          key="2021"
          value={2021}
        >
          2021
        </option>
        <option
          key="2022"
          value={2022}
        >
          2022
        </option>
        <option
          key="2023"
          value={2023}
        >
          2023
        </option>
        <option
          key="2024"
          value={2024}
        >
          2024
        </option>
        <option
          key="2025"
          value={2025}
        >
          2025
        </option>
      </select>
    </div>
    <Button
      className="CalendarHeader__navButton"
      context="neutral"
      disabled={false}
      isBlock={false}
      isInline={false}
      onClick={[MockFunction]}
      size="normal"
      type="button"
    >
      <button
        className="Button Button--context_neutral CalendarHeader__navButton"
        disabled={false}
        onClick={[MockFunction]}
        type="button"
      >
        <ImArrowRight2>
          <IconBase
            attr={
              Object {
                "version": "1.1",
                "viewBox": "0 0 16 16",
              }
            }
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              version="1.1"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9.707 13.707l5-5c0.391-0.39 0.391-1.024 0-1.414l-5-5c-0.391-0.391-1.024-0.391-1.414 0s-0.391 1.024 0 1.414l3.293 3.293h-9.586c-0.552 0-1 0.448-1 1s0.448 1 1 1h9.586l-3.293 3.293c-0.195 0.195-0.293 0.451-0.293 0.707s0.098 0.512 0.293 0.707c0.391 0.391 1.024 0.391 1.414 0z"
                key="0"
              />
            </svg>
          </IconBase>
        </ImArrowRight2>
      </button>
    </Button>
  </div>
</CalendarHeader>
`;
